### ISSUE TITLE
风格特征可视化

### ISSUE THREAD

**Issue by Yuefengxin** [wenyh1616/StyleMotion#2]
@wenyh1616 您好，方便分享一下风格特征可视化的代码吗？我尝试用您提供的模型进行风格特征可视化，但是和论文中的结果差别较大。想知道如何得到论文中下图所示的结果呢？感谢！
<img width="259" alt="wen" src="https://user-images.githubusercontent.com/46143512/144737517-ac9fe7a3-704c-4be1-95c2-1691dd3f51ca.PNG">
以下是我对风格特征的处理：
```python
    def latent_code_visualize(self, iterations):
        def to_float(item):
            if isinstance(item, torch.Tensor):
                item = item.detach().cpu().numpy()
            if isinstance(item, np.ndarray):
                if len(item.reshape(-1)) == 1:
                    item = float(item)
            return item
        with torch.no_grad():

            """latent codes"""  # !!!!! TD: add a separate function, merge with plot_clusters ????t-????????(t-SNE)???

            vis_dicts = {}
            for phase, co_loader, writer in [
                ['train', self.data_loader, self.writer]]:

                vis_dict = None
                # ??key????list
                for t, tcl_data in enumerate(co_loader):
                    vis_codes = {}
                    x = tcl_data["x"].to(self.data_device)  # 100,63,70
                    cond = tcl_data["cond"].to(self.data_device)  # 100, 663, 70
                    vis_codes['style_code'] = self.graph.generate_z(x, cond)
                    vis_codes['meta'] = tcl_data['meta']
                    vis_codes['label'] = tcl_data['label']
                    if vis_dict is None:
                        vis_dict = {}
                        for key, value in vis_codes.items():
                            vis_dict[key] = [value]
                    else:
                        for key, value in vis_codes.items():
                            vis_dict[key].append(value)

                for key, value in vis_dict.items():
                    if key == "meta" or key == "label":
                        num = len(value)
                        vis_dict[key] = [to_float(item) for i in range(num) for item in value[i]]
                    else:
                        vis_dict[key] = torch.cat(vis_dict[key], 0)
                        vis_dict[key] = vis_dict[key].cpu().numpy()
                        vis_dict[key] = to_float(vis_dict[key].reshape(vis_dict[key].shape[0], -1))

                vis_dicts[phase] = vis_dict

            writers = {"train": self.writer}
            from visualization.latent_plot_utils import get_all_plots
            get_all_plots(vis_dicts, os.path.join(self.log_dir, '%08d' % (iterations + 1)), writers, iterations + 1)
```

**Comment by wenyh1616** [comment]
论文中的动作风格隐编码是各个风格下随机的动作sample的plot出来的结果，不会一模一样。




------------------ 原始邮件 ------------------
发件人: Yuefengxin ***@***.***&gt;
发送时间: 2021年12月5日 15:20
收件人: wenyh1616/StyleMotion ***@***.***&gt;
抄送: wenyh1616 ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)





 
@wenyh1616 您好，方便分享一下风格特征可视化的代码吗？我尝试用您提供的模型进行风格特征可视化，但是和论文中的结果差别较大。想知道如何得到论文中下图所示的结果呢？感谢！
 
 以下是我对风格特征的处理：
     def latent_code_visualize(self, iterations):         def to_float(item):             if isinstance(item, torch.Tensor):                 item = item.detach().cpu().numpy()             if isinstance(item, np.ndarray):                 if len(item.reshape(-1)) == 1:                     item = float(item)             return item         with torch.no_grad():             """latent codes"""  # !!!!! TD: add a separate function, merge with plot_clusters ????t-????????(t-SNE)???             vis_dicts = {}             for phase, co_loader, writer in [                 ['train', self.data_loader, self.writer]]:                 vis_dict = None                 # ??key????list                 for t, tcl_data in enumerate(co_loader):                     vis_codes = {}                     x = tcl_data["x"].to(self.data_device)  # 100,63,70                     cond = tcl_data["cond"].to(self.data_device)  # 100, 663, 70                     vis_codes['style_code'] = self.graph.generate_z(x, cond)                     vis_codes['meta'] = tcl_data['meta']                     vis_codes['label'] = tcl_data['label']                     if vis_dict is None:                         vis_dict = {}                         for key, value in vis_codes.items():                             vis_dict[key] = [value]                     else:                         for key, value in vis_codes.items():                             vis_dict[key].append(value)                 for key, value in vis_dict.items():                     if key == "meta" or key == "label":                         num = len(value)                         vis_dict[key] = [to_float(item) for i in range(num) for item in value[i]]                     else:                         vis_dict[key] = torch.cat(vis_dict[key], 0)                         vis_dict[key] = vis_dict[key].cpu().numpy()                         vis_dict[key] = to_float(vis_dict[key].reshape(vis_dict[key].shape[0], -1))                 vis_dicts[phase] = vis_dict             writers = {"train": self.writer}             from visualization.latent_plot_utils import get_all_plots             get_all_plots(vis_dicts, os.path.join(self.log_dir, '%08d' % (iterations + 1)), writers, iterations + 1)
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Yuefengxin** [comment]
感谢回复！不过我的意思是我跑出来的聚类效果比较差，像下面这样，跟论文的聚类效果不一致。
<img width="273" alt="style_code" src="https://user-images.githubusercontent.com/46143512/144739004-ca517cb2-7a2d-4735-a60f-eafbd28726ca.PNG">

**Comment by wenyh1616** [comment]
我们聚类的动作用的都是walking的数据，请问给出的图里面是不是用了所有content的数据？






------------------ 原始邮件 ------------------
发件人: Yuefengxin ***@***.***&gt;
发送时间: 2021年12月5日 16:17
收件人: wenyh1616/StyleMotion ***@***.***&gt;
抄送: wenyh1616 ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)






感谢回复！不过我的意思是我跑出来的聚类效果比较差，像下面这样，跟论文的聚类效果不一致。


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Yuefengxin** [comment]
确实用了全部数据，原来如此，谢谢！

**Comment by Yuefengxin** [comment]
@wenyh1616 不好意思再打扰一下，我想计算一下不可见风格可视化那张图的SCoeff 和 CHI指标，也就是下图这个。
<img width="256" alt="1" src="https://user-images.githubusercontent.com/46143512/145663528-9c447f79-b360-4849-8647-63590fdc619d.PNG">
但是我用您的源码做实验的效果不太好，即便只针对walking数据，不可见风格也没有聚类。以下是我实验得到的风格和对应的内容的可视化图。
<img width="276" alt="2" src="https://user-images.githubusercontent.com/46143512/145663579-f59eac58-606b-45e3-a7a1-b082bda1259b.PNG">
<img width="277" alt="3" src="https://user-images.githubusercontent.com/46143512/145663602-1b153021-2c26-4ac7-88aa-4618628f2985.PNG">
想请问一下您有计算过这个指标吗，可否提供一下结果呢？因为不知道是不是我实验没有做到位，确实不能还原论文中的效果。感谢！

**Comment by wenyh1616** [comment]
您好，当时的实验模型没有全部保留，如果需要的话我还要再跑，目前的plot隐编码的方式有没有问题呢？现在看起来即使是除了strutting之外的运动风格也不能聚类，不太正常。







------------------ 原始邮件 ------------------
发件人: Yuefengxin ***@***.***&gt;
发送时间: 2021年12月11日 12:26
收件人: wenyh1616/StyleMotion ***@***.***&gt;
抄送: wenyh1616 ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)






@wenyh1616 不好意思再打扰一下，我想计算一下不可见风格可视化那张图的SCoeff 和 CHI指标，也就是下图这个。

但是我用您的源码做实验的效果不太好，即便只针对walking数据，不可见风格也没有聚类。以下是我实验得到的风格和对应的内容的可视化图。


想请问一下您有计算过这个指标吗，可否提供一下结果呢？因为不知道是不是我实验没有做到位，确实不能还原论文中的效果。感谢！

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by wenyh1616** [comment]
看起来你发的图并不是只对walking聚类的，也包含了其他类型的运动不是吗？




------------------ 原始邮件 ------------------
发件人: Yuefengxin ***@***.***&gt;
发送时间: 2021年12月11日 12:26
收件人: wenyh1616/StyleMotion ***@***.***&gt;
抄送: wenyh1616 ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)






@wenyh1616 不好意思再打扰一下，我想计算一下不可见风格可视化那张图的SCoeff 和 CHI指标，也就是下图这个。

但是我用您的源码做实验的效果不太好，即便只针对walking数据，不可见风格也没有聚类。以下是我实验得到的风格和对应的内容的可视化图。


想请问一下您有计算过这个指标吗，可否提供一下结果呢？因为不知道是不是我实验没有做到位，确实不能还原论文中的效果。感谢！

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Yuefengxin** [comment]
感谢回复！是包含了其它类型的，不过第二张图中的橙色就是walking的数据，可以对应起来看。我用的plot是直接使用的Aberman这个仓库的代码 https://github.com/DeepMotionEditing/deep-motion-editing/blob/master/style_transfer/probe/latent_plot_utils.py，想来应该没有什么问题。

**Comment by Yuefengxin** [comment]
那请问您方便跑一下unseen的模型吗？因为我最近在做对比实验，需要这样一个数据。如果可以的话十分感谢！如果不方便的话，我看论文发现unseen的图和所有风格一起训练的图十分相似，可以理解为和这个结果是接近的吗？
<img width="260" alt="4" src="https://user-images.githubusercontent.com/46143512/145664161-73fd524b-45d9-466e-ae0c-cfcb2e6ea9a5.PNG">

**Comment by wenyh1616** [comment]
我的意思是用你现在训练的模型，只计算walking类型的运动结果就可以了～你现在用的不是walking。你的图2和图1结合起来看，还是用的所有content的数据。------------------ 原始邮件 ------------------发件人: Yuefengxin  ***@***.***>发送时间: 2021年12月11日 12:48收件人: wenyh1616/StyleMotion ***@***.***>抄送: wenyh1616 ***@***.***>, Mention ***@***.***>主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)
那请问您方便跑一下unseen的模型吗？因为我最近在做对比实验，需要这样一个数据。如果可以的话十分感谢！如果不方便的话，我看论文发现unseen的图和所有风格一起训练的图十分相似，可以理解为和这个结果是接近的吗？


—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/wenyh1616/StyleMotion/issues/2#issuecomment-991459867",
"url": "https://github.com/wenyh1616/StyleMotion/issues/2#issuecomment-991459867",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by Yuefengxin** [comment]
@wenyh1616 我去除walking之后实验得到的结果是这样的，其实就是之前那个图去掉其它类型的结果，还是没有很好的聚类。。。
<img width="274" alt="6" src="https://user-images.githubusercontent.com/46143512/145674839-8fad8f4e-54ea-4fc0-8c2f-5687e3bd58c8.PNG">
另外由于没有只去除strutting风格的数据集，我是在数据集加载的时候去掉strutting风格的，就是获取数据之后像下面这样，这里会有问题吗？
```
        index = train_metas['style'].index('strutting')
        train_data = train_data[:index]
        train_labels = train_labels[:index]
        train_metas['style'] = train_metas['style'][:index]
        train_metas['content'] = train_metas['content'][:index]
```

**Comment by wenyh1616** [comment]
这样写怎么保证:index去掉了strutting？数据是按照类型排序了吗？这个写的有问题，要调试一下。

---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2021年12月11日(周六) 晚上7:37
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)




 
@wenyh1616 我去除walking之后实验得到的结果是这样的，其实就是之前那个图去掉其它类型的结果，还是没有很好的聚类。。。
 
 另外由于没有只去除strutting风格的数据集，我是在数据集加载的时候去掉strutting风格的，就是获取数据之后像下面这样，这里会有问题吗？
         index = train_metas['style'].index('strutting')         train_data = train_data[:index]         train_labels = train_labels[:index]         train_metas['style'] = train_metas['style'][:index]         train_metas['content'] = train_metas['content'][:index]  
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Yuefengxin** [comment]
@wenyh1616 感谢大佬的耐心！我是在调试的时候看到所有strutting风格都在末尾，所以这样做的。。或者您方便提供一下没有strutting的数据集吗？
<img width="304" alt="7" src="https://user-images.githubusercontent.com/46143512/145675371-03f97284-2659-4dde-aab0-f92811cc64d8.PNG">
不过我现在发现即便是用全部风格训练的模型，只测试walking的效果也没有论文中的好诶，看起来有一点诡异。。。
<img width="276" alt="5" src="https://user-images.githubusercontent.com/46143512/145675405-aa36b164-5d92-4b4c-b8e5-c0d2ea0a247a.PNG">
风格特征是像下面这样获取对吗？
```
                    x = tcl_data["x"].to(self.data_device)  # 100,63,70
                    cond = tcl_data["cond"].to(self.data_device)  # 100, 663, 70
                    vis_codes['style_code'] = self.graph.generate_z(x, cond)
```

**Comment by wenyh1616** [comment]
在分析latent code的时候，对于每个运动片段的latent code应该在时间轴上取个平均。------------------ 原始邮件 ------------------发件人: Yuefengxin  ***@***.***>发送时间: 2021年12月11日 19:52收件人: wenyh1616/StyleMotion ***@***.***>抄送: wenyh1616 ***@***.***>, Mention ***@***.***>主题: Re: [wenyh1616/StyleMotion] 风格特征可视化 (Issue #2)
@wenyh1616 感谢大佬的耐心！我是在调试的时候看到所有strutting风格都在末尾，所以这样做的。。或者您方便提供一下没有strutting的数据集吗？

不过我现在发现即便是用全部风格训练的模型，只测试walking的效果也没有论文中的好诶，看起来有一点诡异。。。

风格特征是像下面这样获取对吗？
                    x = tcl_data["x"].to(self.data_device)  # 100,63,70
                    cond = tcl_data["cond"].to(self.data_device)  # 100, 663, 70
                    vis_codes['style_code'] = self.graph.generate_z(x, cond)


—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/wenyh1616/StyleMotion/issues/2#issuecomment-991615111",
"url": "https://github.com/wenyh1616/StyleMotion/issues/2#issuecomment-991615111",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by Yuefengxin** [comment]
@wenyh1616 噢噢问题就在这里，取完平均值后有一致的结果了，十分感谢！
